proposals/2023-06-13_created-timestamp.md

```diff
@@ -0,0 +1,158 @@
+
+## Goals
+
+* Prometheus can collect and store reset timestamps for counters, summaries, and histograms on scrape using common exposition formats.
+* Prometheus will persist the created timestamp across restarts (WAL).
+* Prometheus can perform meaningful rate-like functions even with a single scrape.
+* Prometheus can inject created timestamps for stateless remote write use.
+  * Stateless means here that every streamed batch of samples for certain counters has its created timestamp known.
+
+### Audience
+
+* Users monitoring push-based applications with long push intervals.
+* Users depending on high-precision rates, especially during the initial scrapes of an application.
+* Users that are sending metrics to ingestors that require created timestamp, e.g. Monarch.
+
+## Non-Goals
+
+* Propose changes to Prometheus text exposition format for now. Advanced functionalities like exemplars, native histograms, or created timestamps are not on the short-term roadmap.
+* Elaborate on technical details for rate implementation.
+* Remote Write extensions (should be in different proposals).
+* Support Gauges and Gauge-like metrics (Info, Stateset, Gauge Histograms). None of the exposition formats allows this, thus it’s out of scope.
+
+## How
+
+We propose to append a newly appearing created timestamp as an additional 0-value sample to the TSDB. We call this a “zero injection”.
+
+During a scrape, created timestamps are processed before normal sample values. If a certain metric is not yet present in the TSDB, a “synthetic” sample is appended with a value of 0 and a timestamp equal to the created timestamp found in the scrape. Followed by appending the actual sample with the usual scraped value and timestamp.
+
+![Initial scrape example](../assets/2023-06-13_created-timestamps/initial-scrape-example.png)
+
+On following scrapes, created timestamps are not appended until a scrape finds a created timestamp higher than the latest sample's timestamp, which would trigger the insertion of another sample with a 0-value sample.
+
+![Alt text](../assets/2023-06-13_created-timestamps/followup-scrape-example.png)
+
+This is already enough to improve rate calculations as mentioned in the [Goals](#goals) section, while also persisting such data in WAL (even if not explicit) and although it allows an easier implementation, it comes with few limitations:
+
+* Unable to ingest created timestamps out of TSDB Head range.
+* Unable to help ingestors that require created timestamps as a separate field, as they would need to implement stateful logic to identify injected zeros in a batch of samples.
+
+For those reasons, created timestamps will also be stored as metadata **per series**, following the similar logic used for the zero-injection.
```

Comment:
It is not entirely clear from this sentence why the synthetic samples are still needed if the created timestamps are also stored as metadata. 
Am I reading this correctly that, the created timestamps are therefore stored twice?

Reply:
Yes, they are stored twice

* Injected in TSDB head so we can better rate results without changing interfaces to make metadata available during query time. Also, it would probably have a performance impact if we needed to inject zeros for every query.
* Added to metadata again, to enable remote-write receivers to easily identify created timestamps.

Reply:
Thanks, the first part was the explanation I was looking for. Thanks for writing it!